core/src/main/java/org/mskcc/oncotree/utils/MainTypesUtil.java

```diff
@@ -1,35 +1,43 @@
 package org.mskcc.oncotree.utils;
 
+import com.google.common.collect.Lists;
 import org.mskcc.oncotree.error.InvalidOncoTreeDataException;
 import org.mskcc.oncotree.model.MainType;
 import org.mskcc.oncotree.model.Version;
 
-import java.util.List;
+import java.util.*;
+import org.mskcc.oncotree.model.TumorType;
 
 /**
  * Created by Hongxin on 2/25/16.
  */
 public class MainTypesUtil {
-    /**
-     * return MainType if not exist, create a new one.
-     *
-     * @param name The searched main type name
-     * @return
-     */
-    public static MainType getOrCreateMainType(String name, Version version) throws InvalidOncoTreeDataException {
-        List<MainType> mainTypes = CacheUtil.getMainTypesByVersion(version);
 
-        for (MainType mainType : mainTypes) {
-            if (mainType.getName().equalsIgnoreCase(name)) {
-                return mainType;
+    public static List<MainType> getMainTypesByTumorTypes(Map<String, TumorType> tumorTypes) {
+        Set<MainType> mainTypes = new HashSet<>();
+        // skip the root node, "TISSUE". Just add it's children
+        for (String tumorTypeName : tumorTypes.keySet()) {
+            TumorType tumorType = tumorTypes.get(tumorTypeName);
+            Map<String, TumorType> children = tumorType.getChildren();
+            for (String code : children.keySet()) {
+                addMainTypesToSet(children.get(code), mainTypes);
             }
         }
+        List<MainType> toReturn = new ArrayList<>();
```

Comment:
ArrayList has a constructor where you can pass in a collection ... rather than doing this in two steps.
Lines 26-28 probably can be "return new ArrayList<>(mainTypes)"